File: VB/WinSolution.Module/ClassesForTree.vb
Imports Microsoft.VisualBasic
Imports DevExpress.Xpo
Imports DevExpress.Persistent.Base
Imports DevExpress.Persistent.BaseImpl
Imports System.ComponentModel
Imports DevExpress.Persistent.Base.General

Namespace WinSolution.Module
	<NavigationItem> _
	Public MustInherit Class Category
		Inherits BaseObject
		Implements ITreeNode
		Private name_Renamed As String
		Protected MustOverride ReadOnly Property Parent() As ITreeNode
		Protected MustOverride ReadOnly Property Children() As IBindingList
		Public Sub New(ByVal session As Session)
			MyBase.New(session)
		End Sub
		Public Property Name() As String
			Get
				Return name_Renamed
			End Get
			Set(ByVal value As String)
				name_Renamed = value
				SetPropertyValue("Name", name_Renamed, value)
			End Set
		End Property
		#Region "ITreeNode"
		Private ReadOnly Property ITreeNode_Children() As IBindingList Implements ITreeNode.Children
			Get
				Return Children
			End Get
		End Property
		Private ReadOnly Property ITreeNode_Name() As String Implements ITreeNode.Name
			Get
				Return Name
			End Get
		End Property
		Private ReadOnly Property ITreeNode_Parent() As ITreeNode Implements ITreeNode.Parent
			Get
				Return Parent
			End Get
		End Property
		#End Region
	End Class
	Public Class Level1
		Inherits Category
		Protected Overrides ReadOnly Property Parent() As ITreeNode
			Get
				Return Nothing
			End Get
		End Property
		Protected Overrides ReadOnly Property Children() As IBindingList
			Get
				Return Level2s
			End Get
		End Property
		Public Sub New(ByVal session As Session)
			MyBase.New(session)
		End Sub
		Public Sub New(ByVal session As Session, ByVal name As String)
			MyBase.New(session)
			Me.Name = name
		End Sub
		<Association("Level1-Level2s"), Aggregated> _
		Public ReadOnly Property Level2s() As XPCollection(Of Level2)
			Get
				Return GetCollection(Of Level2)("Level2s")
			End Get
		End Property
	End Class
	Public Class Level2
		Inherits Category
		Private _level1 As Level1
		Protected Overrides ReadOnly Property Parent() As ITreeNode
			Get
				Return _level1
			End Get
		End Property
		Protected Overrides ReadOnly Property Children() As IBindingList
			Get
				Return Level3s
			End Get
		End Property
		Public Sub New(ByVal session As Session)
			MyBase.New(session)
		End Sub
		Public Sub New(ByVal session As Session, ByVal name As String)
			MyBase.New(session)
			Me.Name = name
		End Sub
		<Association("Level1-Level2s")> _
		Public Property Level1() As Level1
			Get
				Return _level1
			End Get
			Set(ByVal value As Level1)
				SetPropertyValue("Level1", _level1, value)
			End Set
		End Property
		<Association("Level2-Level3s"), Aggregated> _
		Public ReadOnly Property Level3s() As XPCollection(Of Level3)
			Get
				Return GetCollection(Of Level3)("Level3s")
			End Get
		End Property
	End Class
	Public Class Level3
		Inherits Category
		Private _level2 As Level2
		Protected Overrides ReadOnly Property Parent() As ITreeNode
			Get
				Return _level2
			End Get
		End Property
		Protected Overrides ReadOnly Property Children() As IBindingList
			Get
				Return New BindingList(Of Object)()
			End Get
		End Property
		Public Sub New(ByVal session As Session)
			MyBase.New(session)
		End Sub
		Public Sub New(ByVal session As Session, ByVal name As String)
			MyBase.New(session)
			Me.Name = name
		End Sub
		<Association("Level2-Level3s")> _
		Public Property Level2() As Level2
			Get
				Return _level2
			End Get
			Set(ByVal value As Level2)
				SetPropertyValue("Level2", _level2, value)
			End Set
		End Property
	End Class
End Namespace

File: CS/WinSolution.Module/ClassesForTree.cs
using DevExpress.Xpo;
using DevExpress.Persistent.Base;
using DevExpress.Persistent.BaseImpl;
using System.ComponentModel;
using DevExpress.Persistent.Base.General;

namespace WinSolution.Module {
    [NavigationItem]
    public abstract class Category : BaseObject, ITreeNode {
        private string name;
        protected abstract ITreeNode Parent {
            get;
        }
        protected abstract IBindingList Children {
            get;
        }
        public Category(Session session) : base(session) { }
        public string Name {
            get {
                return name;
            }
            set {
                name = value;
                SetPropertyValue("Name", ref name, value);
            }
        }
        #region ITreeNode
        IBindingList ITreeNode.Children {
            get {
                return Children;
            }
        }
        string ITreeNode.Name {
            get {
                return Name;
            }
        }
        ITreeNode ITreeNode.Parent {
            get {
                return Parent;
            }
        }
        #endregion
    }
    public class Level1 : Category {
        protected override ITreeNode Parent {
            get {
                return null;
            }
        }
        protected override IBindingList Children {
            get {
                return Level2s;
            }
        }
        public Level1(Session session) : base(session) { }
        public Level1(Session session, string name)
            : base(session) {
            this.Name = name;
        }
        [Association("Level1-Level2s"), Aggregated]
        public XPCollection<Level2> Level2s {
            get {
                return GetCollection<Level2>("Level2s");
            }
        }
    }
    public class Level2 : Category {
        private Level1 _level1;
        protected override ITreeNode Parent {
            get {
                return _level1;
            }
        }
        protected override IBindingList Children {
            get {
                return Level3s;
            }
        }
        public Level2(Session session) : base(session) { }
        public Level2(Session session, string name)
            : base(session) {
            this.Name = name;
        }
        [Association("Level1-Level2s")]
        public Level1 Level1 {
            get {
                return _level1;
            }
            set {
                SetPropertyValue("Level1", ref _level1, value);
            }
        }
        [Association("Level2-Level3s"), Aggregated]
        public XPCollection<Level3> Level3s {
            get {
                return GetCollection<Level3>("Level3s");
            }
        }
    }
    public class Level3 : Category {
        private Level2 _level2;
        protected override ITreeNode Parent {
            get {
                return _level2;
            }
        }
        protected override IBindingList Children {
            get {
                return new BindingList<object>();
            }
        }
        public Level3(Session session) : base(session) { }
        public Level3(Session session, string name)
            : base(session) {
            this.Name = name;
        }
        [Association("Level2-Level3s")]
        public Level2 Level2 {
            get {
                return _level2;
            }
            set {
                SetPropertyValue("Level2", ref _level2, value);
            }
        }
    }
}